### DISCUSSION TITLE
Sending RAW serial data

### DISCUSSION THREAD

**Discussion by TheBoogieKnight** [Jakeler/ble-serial#82] [General]
Hi there, fantastic project and I apologise in advance if this question is obvious and I'm just being dense. I have done multiple searches for this!

I'm trying to interface a Bosch LIDAR pointer with OsX (M1, Ventura). On Windows I was able to just pass it RAW serial data but all the libraries I've seen on OsX (except depreciated ones that don't work) seem to require the UUID framework which the device doesn't use. Is it possible to just send and receive RAW data using this library?

Thank you so much in advance and all the best.

**Comment by Jakeler** [comment]
Hi @TheBoogieKnight, ble-serial itself does not modify or encode the data in any way, it just forwards the bytes from/to the device file.
Not sure though if I understand your requirement fully and this is raw enough for you. Would be nice if you can give more details, like the exact device, maybe documentation and on which layer you want to send "raw" data.
Of course everything still goes through the OS ble stack:
![grafik](https://user-images.githubusercontent.com/13378765/216159028-d0d694f0-b91a-49b3-9aef-1e8037be8fc0.png)
So it is indeed raw from a App standpoint - then it should already work.
If we are looking on LL on the other hand it is not. This adds ATT and also L2CAP headers, so I would not call it raw anymore. For example when writing a plain string:
![grafik](https://user-images.githubusercontent.com/13378765/216163632-0b85ba13-5785-4379-a3ec-33785ea4e0e9.png)

With the handle value here the stack can resolve the service/characteristic UUIDs. That being said, ATT and GATT are part of the ble core specification, it wouldn't be Bluetooth anymore if these parts are not used.

**Comment by TheBoogieKnight** [comment]
Hi Jakeler and thank you so much for the reply, it's massively appreciated.

I was using this library here with the Bosch Lidar (I'm using a PLR30 but others work too):

https://github.com/philipptrenz/BOSCH-GLM-rangefinder

Basically, it connects to the device (after it's been paired in Windows), then sends raw commands using the serial protocol such as:

b'\xC0\x40\x00\xEE' (This measures distance)

Then receives data back using the serial protocol.

**Comment by Jakeler** [comment]
Aha, now it is much clearer, when looking at the [code](https://github.com/philipptrenz/BOSCH-GLM-rangefinder/blob/master/glm100c.py#L52). Keywords are RFCOMM and Serial Port Profile (SPP), that means it uses Bluetooth Classic. Let me explain this with a graphic again:
![grafik](https://user-images.githubusercontent.com/13378765/216476273-9ab519a0-fdcd-4112-9bae-f2af3310e18a.png)
Source: [BTstack Manual](https://bluekitchen-gmbh.com/btstack/v1.0/protocols/)
The left column is Bluetooth LE (BLE, version 4 - 5), which provides GATT as the upper layer. 
On the right is Bluetooth Classic (version 2), which has (among other options) the RFCOMM / SPP profile. 

As you can see, these versions are not more or less raw, protocols are just different. Changes with version 4+ are actually so big that it breaks compatibility completely. Many devices contain therefore 2 controllers (dual mode) connected to the same lower level interface.

To finally answer your initial question: ble-serial focuses on the new bluetooth versions, so you can't use it with this device. Unfortunately I also don't know any good legacy library for modern platforms, it is all deprecated by now, since BT 2.0 was released in 2004. Probably the best backward compatibility has Android, maybe also iOS. Anyway I hope the infos help a bit to understand the topic and maybe finding a solution.
